### ISSUE TITLE
Feature request: Integrate Tesla API and schedule charging

### ISSUE THREAD

**Issue by pgerstbach** [sp4c38/AWattPrice#72]
Since Tesla added charging scheduling to its app, this feature would be a no-brainer:

* Integrate Tesla API, once charging scheduling is reverse engineered
* Based on the existing "cheapest time" (Günstigste Zeit) calculator set charging schedule accordingly

Although I would prefer a slightly different way to configure it: 
1. Set charging limit in app (or in Tesla app)
2. Set time when charging should be completet (latest)
3. App sets the optimal time when charging should be started, based on exchange prices

**Comment by sp4c38** [comment]
That would also be a cool feature, true!
I don't have that much experience with the Tesla API, but would be worth digging into it.
Got to see if anything can be done here.

**Comment by nottheedge** [comment]
If you need a tester let me know. I have a Tesla. But I think it's also necessary to consider the wallbox behind which might control the charging process, too.

**Comment by sp4c38** [comment]
Thanks for the offer 👍.
Yes, that's right with the wallbox. One solution would be to configure the time on the wallbox side but that's pretty unhandy as there are too many wallbox types. So, the best solution would probably be to configure the scheduled time on the Tesla side. The wallbox than needs to be configured so that it provides power at all time - or at least for the certain time range in which the car would be charged - so that the Tesla can decide when it wants to charge.

**Comment by nottheedge** [comment]
Correct. That would be a way how it could work at least for Tesla. I’m not sure if other BEV also provide such an API.

**Comment by sp4c38** [comment]
Can't tell either. Also would get tricky, since I only have a Tesla available on which I could do testing - but maybe for the future worth implementing for other BEVs. 

Okay, so that would definitely be a bigger feature to implement. So I will see if and when I could implement this. But thanks already for your inspiration and ideas, guys.

**Comment by pgerstbach** [comment]
I would start with Tesla API + Tesla Wall Charger (or any other wall charger that does not have its own management enabled). 

It seems the new APIs have already been reverse engineered:
the new comands are set_scheduled_charging, set_scheduled_departure, set_charging_amps
https://tesla-api.timdorr.com/vehicle/commands/charging

Automating this would be so awesome! Of course I would help as a tester (Tesla Model 3). Some time ago I also did programming but I have no experience in Python and Swift. So I guess I would not be of any help.

**Comment by sp4c38** [comment]
Great, thanks for the link!  Yes, so that's the way to go - using the Tesla API to set charging.
It really would be awesome if this would be automated 👍. I'll definitely approach both of you when the feature is implemented and ready for testing - however as already mentioned, I can't quite say when that will be.
